submariner-operator/diagnose-refactor-ocm.md

```diff
@@ -0,0 +1,152 @@
+ effort in Addon and UI. For simplicity reasons this approach is not favored.
+
+### SubmarinerDiagnose CRD
+
+```Go
+// SubmarinerDiagnoseSpec defines the desired configuration to run SubmarinerDiagnose
+type SubmarinerDiagnoseSpec struct {
+All             bool         `json:"all,omitempty"`
+CNI             bool         `json:"CNI,omitempty"`
+Connections     bool         `json:"connections,omitempty"`
+Deployment      bool         `json:"deployment,omitempty"`
+Firewall        bool         `json:"firewall,omitempty"`
+GatherLogs      bool         `json:"gatherLogs,omitempty"`
+K8sVersion      bool         `json:"k8sVersion,omitempty"`
+KubeProxyMode   bool         `json:"kubeProxyMode,omitempty"`
+FirewallOptions FirewallOptions `json:"firewallOptions,omitempty"`
+
+// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+// Important: Run "make manifests" to regenerate code after modifying this file
+// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+}
+
+type FirewallOptions struct {
+InterCluster bool `json:"interCluster,omitempty"`
+IntraCluster bool `json:"intraCluster,omitempty"`
+Metrics      bool `json:"metrics,omitempty"`
+RemoteCluster            string `json:"remoteCluster,omitempty"`
+RemoteK8sAPIServer       string `json:"remoteK8sAPIServer,omitempty"`
+RemoteK8sAPIServerToken  string `json:"remoteK8sAPIServerToken,omitempty"`
+RemoteK8sCA              string `json:"remoteK8sCA,omitempty"`
+RemoteK8sSecret          string `json:"remoteK8sSecret,omitempty"`
+RemoteK8sRemoteNamespace string `json:"remoteK8sRemoteNamespace,omitempty"`
+}
+
+// SubmarinerDiagnoseStatus defines the observed result of SubmarinerDiagnose
+type SubmarinerDiagnoseStatus struct {
+Conditions []metav1.Condition `json:"conditions"`
+K8sVersion string `json:"k8sVersion,omitempty"`
+CNIType string `json:"cniType,omitempty"`
+KubeProxyMode bool `json:"kubeProxyMode,omitempty"`
+ConnectionsStatus ConnectionsStatus `json:"connectionsStatus,omitempty"`
+FirewallStatus FirewallStatus `json:"firewallStatus,omitempty"`
+// DeploymentStatus already captured in SubmarinerStatus, no need to duplicate information.
+// DeploymentStatus SubmarinerStatus `json:"DeploymentStatus,omitempty"`
+}
+
+type ConnectionsStatus struct {
+// TBD in future when UI supports Topology view
+}
+
+type FirewallPortStatus string
+
+const (
+Allowed = "allowed"
+Blocked = "blocked"
+Unknown = "unknown"
+)
+
+type FirewallStatus struct {
+Metrics FirewallPortStatus `json:"metricsStatus,omitempty"`
+VxLanTunnel FirewallPortStatus `json:"vxLanTunnel,omitempty"`
+IPSecTunnel FirewallPortStatus `json:"IPSecTunnel,omitempty"`
+}
+```
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: submariner-diagnose-xyz123
+spec:
+  template:
+    spec:
+      containers:
+      - name: submariner-diagnose-xyz123
+        image: submariner-diagnose
+        command: ["subctl",  "diagnose", "all", "--in-cluster"]
```

Comment:
Is "--in-cluster" some new flag? Cos' `subctl diagnose all` itself runs within the local cluster alone.